### ISSUE TITLE
revisit design of cal-sailing web cam

### ISSUE THREAD

**Issue by jhpoelen** [jhpoelen/cal-sailing-cam#7]
John B. expressed a desire to learn more about the web cam setup behind https://whatskraken.cal-sailing.org .

We were thinking to meet on Sat Jan 29th after lessons at the cal-sailing shed to chat about it.

@MrMinos @cal-sailing please holler if you are interested in joining.

Here's the repositories and web locations that related to the web cam stuff that I know of -  

https://github.com/jhpoelen/cal-sailing-cam (documentation + bash scripts to push images to server)

https://bitbucket.org/jhpoelen/cal-sailing (server-side java/scala/clojure* code, designed for google cloud, that does the management of images and weather info)

https://github.com/cal-sailing/wind2web (scripts to push wind info to server) 

Curious to hear your thoughts on how to organize stuff so that others can figure out how stuff works now. 

* note that the odd mix of JVM languages on the server-side was a bit of a proof-of-concept for me to figure out ways to use different languages in the same app . . . so it is intentionally heterogeneous. I never really bothered to develop the app beyond this experimental phase.

**Comment by jjbongio** [comment]
So as I understand it, the images are sent to one of your servers, which archives them and makes them available to the CSC website.
Do you think it makes sense to do everything on the CSC website server?

**Comment by jjbongio** [comment]
My first-order issue is having all of it on computers belonging to the Club, especially as you are going out of the area. Whether it's in the webserver or something else is less important.

**Comment by jhpoelen** [comment]
Owning all of it sounds like a great plan.

The DNS entry for whatskraken.cal-sailing.org already belongs to the club, or whoever controls the cal-sailing.org domain, so you should be able to instantiate a new version of the sailing cam in google cloud, or whatever environment you choose. I put a nginx instance in front of google cloud to cache images and avoid hitting the non-free tier. 

The cam server is pretty straight forward if you are familiar with java/mvn/gcloud stack. Plenty of tutorials available online. 

Happy to answer remaining after you give it a go.

**Comment by jjbongio** [comment]
After a quick look at this, I'm tempted to do all of the server side stuff in the site web server (Joomla/PHP), in part because it would be integrated but also because it's a programming environment that I know. I'm not keen to learn google cloud programming unless there's a really good reason for implementing it there.

At a high level, it's a relatively simple application, taking feeds from 3 sources that are well-defined.
I spent a lot of time recently dealing with the NOAA forecast XML to pull the weather forecasts into the lesson announcements, so I know how to do that. And my guess is that the current wind data is pretty straight-forward.

Does this make sense?

I think the easiest way to try it out is to have the Raspberry PI code send the HTTP POSTs to 2 addresses each time, one of which is a test site, so we can operate in parallel while I'm getting it working.

Of course, if someone else has a better idea and is willing to implement it, great.

**Comment by jhpoelen** [comment]
All sounds good and I'd encourage you to let others do the work. 

I was super excited to see @MrMinos make contributions, and realized that code is best when shared, just like sailing boats. 

However, if you are in a hurry, I imagine that you can replicate this functionality in no time in your favorite stack. 

As they say:

If you want to go fast, go alone, 
if you want to go far, go together.

**Comment by jjbongio** [comment]
I'm in no hurry and would love for someone else to do this. Just thinking about what to do if that doesn't happen.

**Comment by jhpoelen** [comment]
perhaps we can discuss and float some ideas around on the 29th Jan. Also, I am not planning to turn off anything just yet and others already worked with bits and pieces of the system, so there's room for a gradual transition, just like the cal-sailing.appspot.com -> whatskraken.cal-sailing.org migration.

**Comment by jjbongio** [comment]
My biggest worry is this. The webcams have become a really important part of the Club, especially since the DL White Board camera has been implemented. With your moving away, I worry that some day it will break, and nobody will be able to fix it easily.

So it's not mainly about the work to transition it, but mainly about the ongoing maintenance of it.

Yes, floating ideas on Saturday is a great idea. Looking forward to it. What time do you think? 1:30? Tony said he'd come. Have you heard from others?

**Comment by jhpoelen** [comment]
neat to see how the cams grew from my laziness to bike 40 mins only to find out there was no wind, to an even more useful insight into the state of the club. I really like the doodles (and important messages) broadcasted into the cyberverse via the whiteboard cam. Reminds me of the many attempts emergency departments have taken to implement fancy electronic record keeping only to revert to whiteboards for their responsiveness, easy of use, and reliability. 

Looking forward to the chat next Saturday after lessons, 1:30p would work for me. I am sure others would join if they are interested.

Also, I am sure that the collective wisdom of the club would help keep the something useful up and running, just like we do with many other, even more complex things like dinghies, keel boats, and the fleet of wind surfers. And, also, probably want to keep your old pair of shoes before your try the new ones, so I like your idea to run things in parallel.

**Comment by jjbongio** [comment]
Can you login to the Raspberry PI over the internet, or only at the Clubhouse via the local net?

**Comment by jhpoelen** [comment]
You can find documentation on how to work with the pi at https://github.com/jhpoelen/cal-sailing-cam . You might find the "rebuilding the raspberry pi" interesting.

**Comment by jjbongio** [comment]
Yeah, I looked at that, but it wasn't clear whether I could access the pi from outside the Clubhouse local net.

**Comment by jhpoelen** [comment]
@jjbongio thanks for being rigorous in understanding how the existing cam setup works. 

I learned a lot from rebuilding a pi independently. 

Also, as far as remote control goes: step 7 of https://github.com/jhpoelen/cal-sailing-cam#rebuilding-the-raspberry-pi shows that the pi is executing a script each minute and this script is hosted on github. I've used this to influence/troubleshoot/improve the behavior of the pi remotely.

**Comment by jjbongio** [comment]
Great session today, thanks. I feel quite comfortable that we'll be able to keep this valuable service going.

One question. I was looking through the project, and I couldn't find how you are getting the weather forecast. I'm especially interested because I've been getting frequent timeouts from the NOAA SOAP server I've been using on the website.

**Comment by jjbongio** [comment]
I have followed Jorrit's suggestion and bought a PI. I'll play around with it and then put it into the Clubhouse as a second PI. We can use it to test Cyril's stuff in parallel with current operation.

**Comment by jhpoelen** [comment]
> One question. I was looking through the project, and I couldn't find how you are getting the weather forecast. I'm especially interested because I've been getting frequent timeouts from the NOAA SOAP server I've been using on the website.

You can find unit tests / implementation related to data feeds in place in:

https://bitbucket.org/jhpoelen/cal-sailing/

For example:

https://bitbucket.org/jhpoelen/cal-sailing/src/master/src/test/java/org/calsailing/service/TidePredictionServiceNOAATest.java
https://bitbucket.org/jhpoelen/cal-sailing/src/master/src/test/java/org/calsailing/service/WeatherForecastServiceNWSTest.java

https://bitbucket.org/jhpoelen/cal-sailing/src/master/src/main/java/org/calsailing/service/WeatherForecastServiceNWS.java
https://bitbucket.org/jhpoelen/cal-sailing/src/master/src/main/java/org/calsailing/service/TidePredictionServiceNOAA.java


You might that the NOAA forecast page is a plain ol' web scrape.

**Comment by jjbongio** [comment]
Got it, thanks. A web scrape. I was doing an XML forecast download and parsing it. I think their server  had problems for a few days, but it seems to be working now.

I'm trying to figure out  how to deal with its occasional outages/non-responsiveness. I have a certain format that I want to present, and I don't like the idea of parsing text forecasts.
I'm thinking of just downloading the XML periodically and cacheing it,  hoping that when I need it there's a good copy.

--


	John

**Comment by jjbongio** [comment]
I got my PI, and I'm playing around with it. Pretty cool.

**Comment by jjbongio** [comment]
Jorrit, when you do the curl -F to upload stuff, what do you have on the server side to process the POST? Is it hand-crafted or some utility?

**Comment by jjbongio** [comment]
I have been able to use curl both ways between Joomla and my PI. So I'll wait for Cyril  to finish what he's doing, so we can figure out how to integrate it into the website.

**Comment by jjbongio** [comment]
I didn't see a need to use Avahi. I setup a static IP address (not necessary but useful) for the PI, but also set up the local router as a DNS server. I can login to it as raspberrypi.local from anywhere on my home net.
